aws-iot-thingtype/src/main/java/software/amazon/iot/thingtype/Translator.java

```diff
@@ -46,63 +54,31 @@
-                ProgressEvent.<ResourceModel, CallbackContext>builder()
-                        .resourceModel(model)
-                        .status(OperationStatus.FAILED)
-                        .errorCode(errorCode)
-                        .build();
-        if (errorCode != HandlerErrorCode.InternalFailure) {
-            progressEvent.setMessage(e.getMessage());
-        }
-        return progressEvent;
-    }
-
 
-    static HandlerErrorCode translateExceptionToErrorCode(Exception e, Logger logger) {
-        logger.log(String.format("Translating exception \"%s\", stack trace: %s",
-                e.getMessage(), ExceptionUtils.getStackTrace(e)));
-
-        // We're handling all the exceptions documented in API docs
-        // https://docs.aws.amazon.com/iot/latest/apireference/API_CreateThingType.html#API_CreateThingType_Errors
-        // (+same pages for other APIs)
-        // For Throttling and InternalFailure, we want CloudFormation to retry, and it will do so based on the error code.
-        // Reference with Retryable/Terminal in comments for each: https://tinyurl.com/y378qdno
+    static BaseHandlerException translateIotExceptionToHandlerException(
+            String resourceIdentifier, String operationName, IotException e) {
         if (e instanceof ResourceAlreadyExistsException) {
-            return HandlerErrorCode.AlreadyExists;
-        } else if (e instanceof software.amazon.cloudformation.exceptions.ResourceAlreadyExistsException) {
-            return HandlerErrorCode.AlreadyExists;
-        } else if (e instanceof InvalidRequestException) {
-            return HandlerErrorCode.InvalidRequest;
-        } else if (e instanceof LimitExceededException) {
-            return HandlerErrorCode.ServiceLimitExceeded;
+            return new CfnAlreadyExistsException(ResourceModel.TYPE_NAME, resourceIdentifier, e);
+        } else if (e instanceof ResourceNotFoundException) {
+            return new CfnNotFoundException(ResourceModel.TYPE_NAME, resourceIdentifier, e);
         } else if (e instanceof UnauthorizedException) {
-            return HandlerErrorCode.AccessDenied;
+            return new CfnAccessDeniedException(e);
         } else if (e instanceof InternalFailureException) {
-            return HandlerErrorCode.ServiceInternalError;
-        } else if (e instanceof InternalException) {
-            return HandlerErrorCode.ServiceInternalError;
+            return new CfnInternalFailureException(e);
         } else if (e instanceof ServiceUnavailableException) {
-            return HandlerErrorCode.ServiceInternalError;
+            return new CfnGeneralServiceException(operationName, e);
+        } else if (e instanceof InvalidRequestException || e instanceof InvalidQueryException) {
+            return new CfnInvalidRequestException(e);
+        } else if (e instanceof LimitExceededException) {
+            return new CfnServiceLimitExceededException(ResourceModel.TYPE_NAME, e.getMessage());
```

Comment:
whats the difference between LimitExceededException and ThrottlingException (L76)

Reply:
LimitExceededException is thrown when a service limit - such as number of tags is exceeded. 
ThrottlingException is more relating to rate of requests

https://docs.aws.amazon.com/iot/latest/apireference/API_TagResource.html